### PR TITLE
JENKINS-36377 Additional APIs for Blue Ocean autofavorite feature

### PR DESCRIPTION
**Pull request by i386**
Needed for Blue Ocean's autofavorite feature. See [JENKINS-36377](https://issues.jenkins-ci.org/browse/JENKINS-36377).

Changes include:
* Add API to check if user has ever set a favourite for a job
* Add API to toggle favourite
* Bump base POM
* Fix findbugs issues

PTAL @imeredith @larrys

### CODE REVIEW

**Comment by michaelneale** [comment]
nice

File: pom.xml [hunk 1]
```diff
@@ -46,6 +46,16 @@
     <dependencies>
         <dependency>
             <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>mailer</artifactId>
+            <version>1.18</version>
+        </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>matrix-project</artifactId>
```
Comment (michaelneale) [inline pom.xml]: matrix is already bundled in core, is there something here that uses a class from it?

Comment (michaelneale) [inline pom.xml]: 👍 

Yes it came out of core into a plugin at some point.

**Comment by larrys** [comment]
:+1:

File: src/main/java/hudson/plugins/favorite/user/FavoriteUserProperty.java [hunk 2]
```diff
@@ -53,12 +58,37 @@ public UserProperty reconfigure(StaplerRequest req, JSONObject form) throws Desc
...
     @Override
     public UserPropertyDescriptor getDescriptor() {
         return DESCRIPTOR;
     }
 
+    /**
+     * Migrates this properties storage from favourite's list to a map of booleans
+     * @return this
+     */
+    Object readResolve() {
```
Comment (i386) [inline src/main/java/hudson/plugins/favorite/user/FavoriteUserProperty.java]: @larrys I believe this should migrate the data structure from a list to a map.

**Comment by michaelneale** [comment]
@i386 you know have commit access :shipit:

File: src/main/java/hudson/plugins/favorite/Favorites.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+import hudson.model.User;
+import hudson.plugins.favorite.user.FavoriteUserProperty;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+
+/**
+ * Public API for Favorites
+ */
+public final class Favorites {
```
Comment (i386) [inline src/main/java/hudson/plugins/favorite/Favorites.java]: @michaelneale overnight I wrote a nicer public API for Favorites that hides all the rubbish you need todo with the user property.

File: src/test/java/hudson/plugins/favorite/user/FavoritesTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import java.io.IOException;
+
+import static junit.framework.Assert.assertTrue;
+import static junit.framework.Assert.fail;
+import static junit.framework.TestCase.assertFalse;
+
+public class FavoritesTest {
```
Comment (i386) [inline src/test/java/hudson/plugins/favorite/user/FavoritesTest.java]: @michaelneale now comes with better test coverage!

Comment (michaelneale) [inline src/test/java/hudson/plugins/favorite/user/FavoritesTest.java]: 💯

**Pull request merged**